### PR TITLE
Add multiclip loading and reward

### PR DESCRIPTION
**Pull request by tfederico**
I tried to implement the multi-clip reward described in the DeepMimic paper.
In this pull request you will find:

- loading of multiple clips and multiple kinematic characters
- downsampling (using random sampling) and upsampling (using SLERP) of clips in case of multiple clips with different duration
- selection of the maximum reward among all the clips, as described in the paper
- `train` and `run` files to use as arguments
- motions file to try the training and evaluation
- a script to visualize the train and test return from `output/agent0_log.txt`

I chose create new files as I am not 100% sure about the implementation. At the moment, I managed to train the character with multiple walking clips, but it limps. However, if a single clip is provided, it works exactly as the current version of `train_humanoid3d_walk.txt` and all the others single clip instances.

I am putting this out there so that someone else can try it out as well. The only significant difference is that in the configuration file, instead of indicating a file containing the motion data, you can indicate the folder containing the motion data.

### CODE REVIEW

**Comment by tfederico** [comment]
Here is an [example](https://drive.google.com/file/d/1QnWDl9k9QIxauE9h4NTgNyXYGQXa1yCS/view?usp=sharing) of the rewards for the multi-clip training with the uploaded motion files

**Comment by tfederico** [comment]
I sorted out the limping problem by removing one of the motion policies (I guess having 3 "turn right" and 2 "turn left" created some sort of imbalance), now it uses 5 motion files as described in the paper.

**Comment by ManifoldFR** [comment]
Interesting! Would you be interested in making a Gym wrapper for this? I contributed the Gym interface for the base single-clip env a few months back

**Comment by tfederico** [comment]
I can give it a shot when I have time but have honestly no idea about how to do it. Do you have any pointers?

**Comment by ManifoldFR** [comment]
The file I edited was [this one](https://github.com/bulletphysics/bullet3/blob/master/examples/pybullet/gym/pybullet_envs/deep_mimic/gym_env/deep_mimic_env.py). I think it could be re-used as-is, pretty much

**Comment by tfederico** [comment]
@ManifoldFR I took a look at you gym wrapper and tried to run it, but it seems that the simulated character is "flying away" at the beginning of each run. Could you confirm this behaviour? I think that to see that you need to change `renders` in `deep_mimic_env.py` to `True`.

**Comment by ManifoldFR** [comment]
Which agent are you running ? Are you sampling random actions uniformly in
the action space ?

**Comment by tfederico** [comment]
I run the `deepmimic_test.py` file in the `gym_env` folder

**Comment by ManifoldFR** [comment]
Oh yeah, I haven't updated that file. It is actually uniformly sampling
actions in the action space. Try modifying it and sampling truncated
Gaussians or something. Uniform actions are very noisy and uncorrelated so
the character's inertia is going to be all over the place; hence the
character flies off into space.

**Comment by ManifoldFR** [comment]
Also, the Gym env's action space isn't the effective action space since
they are (optionally, but by default) rescaled using the action offset and
scale parameters as in Jason Peng's initial code. I have a fix for the
action space def in my fork of Bullet.

**Comment by tfederico** [comment]
How did you sample uniformly in the action space? Did you define a custom Box space and overwrite the sample method?

**Comment by ManifoldFR** [comment]
Oh, the sample method on the Box class already samples uniformly

**Comment by tfederico** [comment]
sorry, I meant sampling truncated Gaussians! My bad

**Comment by ManifoldFR** [comment]
Oh, you can use scipy.stats' truncnorm with the Box's lower and higher
bounds

**Comment by tfederico** [comment]
@ManifoldFR have you actually managed to train an agent using the gym env? Because I tried several times but I didn't manage to achieve anything that was comparable to the original results

**Comment by ManifoldFR** [comment]
Right now my current setup is I'm using the PPO from `stable-baselines3` with the following options for the model kwargs:
```yaml
DeepMimic:
  norm_reward: false
  learning_rate: 2.5e-6
  n_epochs: 4
  target_kl: 0.05
  ent_coef: 0
  vf_coef: 5.
  max_grad_norm: 100.
```
and the following policy kwargs:
```python
        weight_decay = 1.0e-5
        policy_kwargs = dict(
            activation_fn=nn.ReLU,
            net_arch=[dict(pi=[1024, 512], vf=[1024, 512])],
            log_std_init=log_std_init,
            ortho_init=True,
            optimizer_kwargs=dict(weight_decay=weight_decay)
        )
```

Look at the rl-baselines-zoo repo for the training script: https://github.com/DLR-RM/rl-baselines3-zoo

I had a trained policy a few months back for the backflip motion: https://www.youtube.com/watch?v=K-fYcoDdP7M

**Comment by tfederico** [comment]
Thanks, I will try that. I tried using PPO1 from `stable-baselines` but it would always yield to poor results. I think there are too many differences when compared to the original code, even though they managed to use it for [motion imitation](https://github.com/google-research/motion_imitation/tree/d11f131f2d7ddd9db5ee3335b60af8e6749053f9).

**Comment by ManifoldFR** [comment]
If I recall correctly, they subclassed PPO1 and made a few modifications in
the motion imitation repo

**Comment by tfederico** [comment]
Yes they did something like that. What's the value of `log_std_init`?

**Comment by erwincoumans** [comment]
This way to train should work, it is not using the untested HumanoidDeepMimicWalkBulletEnv environment:
~~~~~~~~~~~~~
pip3 install -U pybullet
python3 -m pybullet_envs.deep_mimic.testrl --arg_file run_humanoid3d_walk_args.txt
~~~~~~~~~~~~~

Here is a way to run a trained policy:

pip3 install -U pybullet
python3 -m pybullet_envs.deep_mimic.testrl --arg_file run_humanoid3d_walk_args.txt

This should be very similar to Jason Peng's original implementation.

Another option is using the motion_imitation environment, it uses baselines to train.

If someone makes the HumanoidDeepMimicWalkBulletEnv gym environment work, please let us know.

**Comment by erwincoumans** [comment]
@tfederico  What is the status of this Pull Request?

@ManifoldFR do you mind creating a pull request so the HumanoidDeepMimicWalkBulletEnv and/or HumanoidDeepMimicBackflipBulletEnv works?

This way to train should work, it is not using the untested HumanoidDeepMimicWalkBulletEnv environment:
~~~~~~~~~~~~~
pip3 install -U pybullet
python3 -m pybullet_envs.deep_mimic.testrl --arg_file run_humanoid3d_walk_args.txt
~~~~~~~~~~~~~

Here is a way to run a trained policy:
~~~~~~~~~~~~~
pip3 install -U pybullet
python3 -m pybullet_envs.deep_mimic.testrl --arg_file run_humanoid3d_walk_args.txt
~~~~~~~~~~~~~
This should be very similar to Jason Peng's original implementation.

Another option is using the motion_imitation environment, it uses baselines to train.

If someone makes the HumanoidDeepMimicWalkBulletEnv gym environment work, please let us know.

**Comment by tfederico** [comment]
@erwincoumans it includes the multiclip reward, but there is no goal at the moment so it's not easy to evaluate. I trained it on multiple walking clips and it works.

**Comment by tfederico** [comment]
If anyone manages to add that, I think the `*_multiclip.py` files can substitute the original ones (e.g., `testrl`, `mpi_run`)

**Comment by ManifoldFR** [comment]
> Yes they did something like that. What's the value of `log_std_init`?

Whoops sorry, it's -3, which gives a std init of about 0.05.

@erwincoumans I did make the `HumanoidDeepMimicWalkBulletEnv` Gym env work, it was a PR I made a few months back. I can make a PR again with a few other modifications though (namely the camera), but they go further than modifying just the Gym wrapper (it includes the COM reward and modifying how the end effector reward is computed).

**Comment by erwincoumans** [comment]
> I trained it on multiple walking clips and it works.

How did you train it exactly, can you provide a full command-line of train and also 'enjoy'/inspect the results?

Thanks!

**Comment by erwincoumans** [comment]
> I can make a PR again with a few other modifications though (namely the camera), but they go further than modifying just the Gym wrapper

Yes, please can a new PR. Would it be compatible with this PR?

**Comment by tfederico** [comment]
> > I trained it on multiple walking clips and it works.
> 
> How did you train it exactly, can you provide a full command-line of train and also 'enjoy'/inspect the results?
> 
> Thanks!

`python3 mpi_run_multiclip.py --num_workers 8 --arg_file train_humanoid3d_walker_args.txt` for training and 
`python3 testrl_multiclip.py --arg_file run_humanoid3d_walker_args.txt` for enjoy.

It's almost identical to the walk clip, except that the argument file has a folder instead of a single file as input.

In case of clips with different durations, you can use the flag `self._downsample` to perform down/upsampling (the formed using random uniform downsampling and the latter using SLERP). The only problem with this is that it probably changed the motion speed, so we should probably look into that.

**Comment by erwincoumans** [comment]
Thanks @tfederico, let's merge it.

**Comment by tfederico** [comment]
@ManifoldFR am I right to assume that you hyperparameters file looks something like this?

```
HumanoidDeepMimicWalkBulletEnv-v1:
  normalize: "dict(norm_reward=False)"
  n_envs: 8
  learning_rate: 2.5e-6
  n_timesteps: !!float 6e7
  policy: 'MlpPolicy'
  n_epochs: 4
  target_kl: 0.05
  ent_coef: 0
  vf_coef: 5.
  max_grad_norm: 100.
  policy_kwargs: "dict(
    activation_fn=nn.ReLU,
    net_arch=[dict(pi=[1024, 512], vf=[1024, 512])],
    log_std_init=-3,
    ortho_init=True,
    optimizer_kwargs=dict(weight_decay=1.0e-5)
			)"
```

**Comment by ManifoldFR** [comment]
That's right! I actually load some of the keys within Python but that's
what it boils down to.

**Comment by tfederico** [comment]
I tried training it with a custom script 

```
n_envs = 8
env = make_vec_env('HumanoidDeepMimicWalkBulletEnv-v1', n_envs=n_envs)
env = VecNormalize(env, norm_reward=model_args['norm_reward'])

model = PPO(
    'MlpPolicy',
    env,
    learning_rate=model_args['learning_rate'],
    n_epochs=model_args['n_epochs'],
    ent_coef=model_args['ent_coef'],
    vf_coef=model_args['vf_coef'],
    max_grad_norm=model_args['max_grad_norm'],
    target_kl=model_args['target_kl'],
    policy_kwargs=policy_kwargs
)

n_steps = 60000000

model.learn(n_steps)
```

but when I tried to load the saved model and run a few steps it wouldn't last longer than ~20/30 steps. I will try to use the training script from the zoo as you suggested. 

Are you aware of any major differences between their script and what I implemented?

**Comment by ManifoldFR** [comment]
Did you run the learned model in deterministic mode? Also be sure to normalize observations too (`norm_obs=True`) during training, that's usually good with continuous observation space control problems.

The training script from the zoo is slightly different from mine, there's a "normalize" key in the yaml and setting it to true, sets normalize both obs and reward to true.

**Comment by tfederico** [comment]
That's how I load and run the deterministic model:
```
env = make_vec_env('HumanoidDeepMimicWalkBulletEnv-v1')
env = VecNormalize(env, training=False, norm_reward=False)

model = PPO.load("/tmp/gym/best_model.zip", env=env)

avg, std = evaluate_policy(model, env, n_eval_episodes=10, deterministic=True)
```
`norm_obs` should be `True` by default.

**Comment by tfederico** [comment]
During training the reward increases up to ~300, e.g.

```
Eval num_timesteps=34000000, episode_reward=270.24 +/- 14.27
Episode length: 578.40 +/- 11.54
```

But then at test time it's really low.

**Comment by ManifoldFR** [comment]
Oh, you should load the VecNormalize checkpoint, too, as `env = VecNormalize.load(path)`. I think the training script is supposed to dump a `vecnormalize.pkl` which contains the saved mean and stdev.

**Comment by ManifoldFR** [comment]
@erwincoumans I made a PR for my new changes, it's gonna be WIP for now. Ideally I'd like to add a Gym env for @tfederico new multiclip env introduced with this PR.
